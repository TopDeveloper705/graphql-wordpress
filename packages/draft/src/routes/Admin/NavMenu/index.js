import React, { Component, Fragment } from 'react';
import { withRouter } from 'react-router';
import {
  Nav,
  navFoldedClass,
  NavLink,
  activeClass,
  foldedActiveClass,
  SubNav,
  SubNavLink,
  subNavActiveClass,
  Separator,
  Dashicon,
  CollapseButton,
  CollapseButtonIcon,
  CollapseButtonLabel,
} from './styled';

/* eslint-disable react/prop-types */

@withRouter
export default class NavMenu extends Component {
  onClick = e => {
    e.preventDefault();

    this.props.onFolded(!this.props.folded);
  };

  render() {
    const { location, routeConfig, folded } = this.props;

    return (
      <Nav className={folded && navFoldedClass}>
        {routeConfig.map((items, i) => (
          <Fragment key={i.toString(16)}>
            {i > 0 && <Separator />}
            {items.map((item, j) => (
              <Fragment key={j.toString(16)}>
                <NavLink
                  to={item.path}
                  exact={item.path === '/'}
                  activeClassName={`${activeClass} ${this.props.folded && foldedActiveClass}`}
                >
                  {item.dashicon && (
                    <Dashicon className={`dashicons-before dashicons-${item.dashicon}`} />
                  )}
                  {!this.props.folded && item.label}
                </NavLink>
                {!folded &&
                  location.pathname.indexOf(item.path) === 0 &&
                  item.routes && (
                    <SubNav key="subnav">
                      {item.routes.map(route => (
                        <SubNavLink
                          key={route.path}
                          to={route.path}
                          exact
                          activeClassName={subNavActiveClass}
                        >
                          {route.label}
                        </SubNavLink>
                      ))}
                    </SubNav>
                  )}
              </Fragment>
            ))}
          </Fragment>
        ))}
        <Separator />
        <CollapseButton folded={this.props.folded} onClick={this.onClick}>
          <CollapseButtonIcon />
          {!this.props.folded && <CollapseButtonLabel>Collapse menu</CollapseButtonLabel>}
        </CollapseButton>
      </Nav>
    );
  }
}
