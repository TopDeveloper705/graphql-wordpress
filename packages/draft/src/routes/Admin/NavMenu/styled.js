import styled from 'react-emotion';
import { css } from 'emotion';

export const Nav = styled.nav`
  background-color: ${p => p.theme.colors.background};
  bottom: -120px;
  height: 100%;
  left: 0;
  position: fixed;
  top: 0;
  width: 160px;
  z-index: 4;

  @media screen and (max-width: 782px) {
    width: 36px;
  }
`;

export const navCollapsedClass = css`
  width: 36px;
`;

export const Separator = styled.i`
  display: block;
  height: 5px;
  margin: 0 0 6px 0;
`;

export const NavWrap = styled.div`
  position: relative;
`;
