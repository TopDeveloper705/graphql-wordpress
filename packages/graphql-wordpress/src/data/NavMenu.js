import { toGlobalId } from 'graphql-relay';

class NavMenu {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return process.env.WP_NAV_MENUS_ENDPOINT || 'graphql/v1/nav-menus';
  }
}

export default NavMenu;
