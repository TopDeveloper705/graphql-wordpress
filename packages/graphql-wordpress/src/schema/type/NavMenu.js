import { GraphQLObjectType, GraphQLList } from 'graphql';
import NavMenuItemType from 'type/NavMenuItem';
import { globalIdField, name } from 'field/identifier';
import description from 'field/description';
import { registerNodeType, NodeInterface } from 'type/relayNode';

const NavMenuType = new GraphQLObjectType({
  name: 'NavMenu',
  description: 'A nav menu.',
  interfaces: [NodeInterface],
  fields: {
    id: globalIdField(),
    ...name,
    ...description,
    items: {
      type: new GraphQLList(NavMenuItemType),
      description: 'Items associated with the menu.',
    },
  },
});

registerNodeType(NavMenuType);

export default NavMenuType;
