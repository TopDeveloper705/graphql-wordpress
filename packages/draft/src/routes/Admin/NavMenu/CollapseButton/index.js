import React from 'react';
import { withTheme } from 'emotion-theming';
import { CollapseButton as StyledButton, buttonIconClass, buttonLabelClass } from './styled';

/* eslint-disable react/prop-types */

function CollapseButton({ theme: { isCollapsed }, ...rest }) {
  return (
    <StyledButton
      {...rest}
      aria-expanded={!isCollapsed}
      aria-label={isCollapsed ? 'Expand menu' : 'Collapse menu'}
    >
      <i className={buttonIconClass} />
      {!isCollapsed && <i className={buttonLabelClass}>Collapse menu</i>}
    </StyledButton>
  );
}

export default withTheme(CollapseButton);
