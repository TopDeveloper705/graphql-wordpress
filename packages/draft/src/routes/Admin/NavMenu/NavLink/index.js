import React from 'react';
import { cx } from 'emotion';
import { NavLink as StyledNavLink, activeClass, Dashicon } from './styled';

/* eslint-disable react/prop-types */

export default function NavLink({ item, hovered, collapsed }) {
  return (
    <StyledNavLink
      to={item.path}
      exact={item.path === '/'}
      activeClassName={cx('NavLink-active', activeClass)}
      className={cx({
        'NavLink-collapsed': collapsed,
        'NavLink-hovered': hovered,
        'NavLink-hasSubNav': item.routes && item.routes.length > 0,
      })}
    >
      {item.dashicon && <Dashicon className={`dashicons-before dashicons-${item.dashicon}`} />}
      <span>{!collapsed && item.label}</span>
    </StyledNavLink>
  );
}
