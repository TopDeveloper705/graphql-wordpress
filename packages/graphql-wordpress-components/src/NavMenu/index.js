import url from 'url';
import React, { Component } from 'react';
import styled from 'react-emotion';
import { Link } from 'found';
import responsive from '../responsive';
import { sortOrderedHierarchy } from '../walker';

export const NavWrapper = styled.nav`
  display: block;
  margin: ${p => p.theme.padding / 2}px 0;

  & li {
    position: relative;
  }

  & a {
    color: ${p => p.theme.colors.black};
    display: block;
    font-size: 14px;
    line-height: ${p => p.theme.padding * 1.5}px;
    padding: 0 ${p => p.theme.padding * 0.75}px;
    text-decoration: none;
    text-transform: uppercase;

    ${responsive.desktop} {
      font-size: 14px;
      line-height: ${p => p.theme.padding * 2}px;
      padding: 0 ${p => p.theme.padding}px;
    }
  }

  & li:hover > a,
  & ul ul :hover > a,
  & a:focus {
    background: ${p => p.theme.colors.subnav.hoverBackground};
  }

  & ul li:hover > ul {
    display: block;
  }
`;

export const Level = styled.ul`
  list-style: none;
  margin: 0 0 0 ${p => p.theme.padding / 2 * -1}px;

  & ul {
    display: none;
    left: ${p => p.theme.padding}px;
    position: absolute;
    top: ${p => p.theme.padding * 1.5}px;
    z-index: 99999;

    ${responsive.desktop} {
      top: ${p => p.theme.padding * 2}px;
    }

    & ul {
      left: 100%;
      top: 0;
    }
  }
`;

export const NavItem = styled.li`
  display: inline-block;
`;

export const SubNavItem = styled(Link)`
  background: ${p => p.theme.colors.subnav.background};
  border-bottom: 1px dotted ${p => p.theme.colors.subnav.detail};
  color: ${p => p.theme.colors.dark};
  display: block;
  font-size: 13px;
  height: auto;
  line-height: 20px;
  padding: ${p => p.theme.padding / 2}px ${p => p.theme.padding}px;
  width: 168px;
`;

type NavMenuItem = {
  id: string,
  title: string,
  url: string,
  parent: string,
  order: number,
  type: string,
  typeName: string,
  typeSlug: string,
  dataSlug: string,
  dataID: string,
};

type NavMenuProps = {
  navMenu: {
    id: string,
    name: string,
    items: Array<NavMenuItem>,
  },
};

export default class NavMenu extends Component {
  props: NavMenuProps;

  sorted = null;

  parseItem = ({ id, parent, title, url: itemUrl, typeSlug, dataSlug }) => {
    let path;
    if (typeSlug && dataSlug) {
      path = `/${typeSlug}/${dataSlug}`;
    } else {
      const urlObj = url.parse(itemUrl);
      if (urlObj.path === '/') {
        path = urlObj.path;
      } else {
        path = urlObj.path.replace(/\/$/, '');
      }
    }

    return (
      <NavItem key={id}>
        {parent ? <SubNavItem to={path}>{title}</SubNavItem> : <Link to={path}>{title}</Link>}
        {this.sorted[id] ? this.walk(this.sorted[id]) : null}
      </NavItem>
    );
  };

  walk(node) {
    return <Level>{node.map(this.parseItem)}</Level>;
  }

  render() {
    const { navMenu } = this.props;

    if (!navMenu) {
      return null;
    }

    this.sorted = sortOrderedHierarchy(navMenu.items);
    const navMenuHtml = this.walk(this.sorted.top);

    return <NavWrapper>{navMenuHtml}</NavWrapper>;
  }
}
