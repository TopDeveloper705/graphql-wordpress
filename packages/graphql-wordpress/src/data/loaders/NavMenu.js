import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import NavMenu from 'data/NavMenu';

// Dataloader expects IDs that can be read by the REST API

export default function getNavMenuLoaders() {
  const endpoint = NavMenu.getEndpoint();

  const navMenuLoader = new Dataloader(ids =>
    fetchData(endpoint)
      .then(({ data: { body } }) => body)
      .then(menus => ids.map(id => menus.find(item => item.id === parseInt(id, 10))))
  );

  return {
    load: async id => {
      const data = await navMenuLoader.load(id);
      return data ? Object.assign(new NavMenu(), data) : null;
    },
  };
}
