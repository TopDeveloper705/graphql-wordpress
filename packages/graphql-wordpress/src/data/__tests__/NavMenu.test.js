import NavMenu from 'data/NavMenu';

const backupEnv = Object.assign({}, process.env);

afterAll(() => {
  process.env = backupEnv;
});

describe('Test NavMenu data access', () => {
  test('Get endpoint', () => {
    expect(NavMenu.getEndpoint()).toMatchSnapshot();
  });

  test('Get backup endpoint', () => {
    delete process.env.WP_NAV_MENUS_ENDPOINT;
    expect(NavMenu.getEndpoint()).toMatchSnapshot();
  });
});
