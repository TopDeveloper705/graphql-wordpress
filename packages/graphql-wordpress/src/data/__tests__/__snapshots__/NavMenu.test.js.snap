// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NavMenu data access Get backup endpoint 1`] = `"graphql/v1/nav-menus"`;

exports[`Test NavMenu data access Get endpoint 1`] = `"graphql/v1/nav-menus"`;
